libvast/builtins/endpoints/query.cpp

```diff
@@ -181,13 +150,68 @@ struct query_manager_state {
   std::optional<pipeline_executor> pipeline_executor_;
 
   std::string create_response() {
-    auto [drained_events, events_as_json]
-      = drain_buffer(processed_slices, limit);
-    shippable_events_count -= drained_events;
-    shipped_events_count += drained_events;
-    auto events = detail::split(events_as_json, "\n");
-    return fmt::format("{{\"position\": {}, \"events\": [\n{}]}}\n",
-                       shipped_events_count, fmt::join(events, ",\n"));
+    auto printer = json_printer{{.oneline = true}};
+    auto result = std::string{};
+    auto out_iter = std::back_inserter(result);
+    auto seen_schemas = std::unordered_set<type>{};
+    auto written = size_t{0};
+    out_iter = fmt::format_to(out_iter, "{{\"events\":[");
+    // write slices
+    auto it = processed_slices.begin();
+    for (bool first = true; it != processed_slices.end(); ++it) {
+      auto slice = std::move(*it);
+      if (slice.rows() == 0)
+        continue;
+      if (const auto remaining = limit - written; slice.rows() > remaining) {
+        auto [head, tail] = split(slice, remaining);
+        *it = std::move(tail);
+        slice = std::move(head);
+      }
+      seen_schemas.insert(slice.schema());
+      auto resolved_slice = resolve_enumerations(slice);
+      auto type = caf::get<record_type>(resolved_slice.schema());
+      auto array
+        = to_record_batch(resolved_slice)->ToStructArray().ValueOrDie();
+      for (const auto& row : values(type, *array)) {
+        if (first)
+          out_iter = fmt::format_to(out_iter, "{{");
+        else
+          out_iter = fmt::format_to(out_iter, "}},{{");
+        first = false;
+        out_iter = fmt::format_to(
+          out_iter, "\"schema-ref\":\"{:x}\",\"data\":", hash(slice.schema()));
+        VAST_ASSERT_CHEAP(row);
+        const auto ok = printer.print(out_iter, *row);
+        VAST_ASSERT_CHEAP(ok);
+      }
+      written += slice.rows();
+      if (written >= limit) {
+        break;
+      }
+    }
+    //  Remove events that will be shipped in the response.
+    processed_slices.erase(processed_slices.begin(), it);
+    // Write schemas
+    if (written == 0) {
+      out_iter = fmt::format_to(out_iter, "],\"schemas\":[]}}\n");
+      return result;
+    }
+    out_iter = fmt::format_to(out_iter, "}}],\"schemas\":[");
+    for (bool first = true; const auto& schema : seen_schemas) {
+      if (first)
+        out_iter = fmt::format_to(out_iter, "{{");
+      else
+        out_iter = fmt::format_to(out_iter, "}},{{");
+      first = false;
+      out_iter = fmt::format_to(
+        out_iter, "\"schema-ref\":\"{:x}\",\"definition\":", hash(schema));
+      const auto ok = printer.print(out_iter, schema.to_definition(true));
+      VAST_ASSERT_CHEAP(ok);
+    }
+    out_iter = fmt::format_to(out_iter, "}}]}}\n");
+    shippable_events_count -= written;
+    shipped_events_count += written;
```

Comment:
I see that you have removed the "position" from the response.
With that being gone we can remove shipped_events_count completely as it is now not used elsewhere

Reply:
Good point; I removed the variable in https://github.com/tenzir/vast/pull/2952/commits/4750258412fa7a4dfeedc30ea798f61b2a88982f